### DISCUSSION TITLE
Phrase construction or enable_encoder/table_translator documentation?

### DISCUSSION THREAD

**Discussion by erikli** [rime/home#1082] [Q&A] [answered]
I want to know how full codes for phrases get converted to shorter codes and couldn't figure it out after trying to read https://github.com/rime/home/wiki/RimeWithSchemata. Is this explained somewhere?

for example in cangjie5:
身體健康 is originally 竹難竹、月月廿田廿、人弓大手、戈中水 and is shortened to become 竹田人手水 and I really don't understand why that happens.

What are the rules? Thank you for any help!

**Comment by erikli** [comment]
雖然用英文回答會讓我多容易明白，用中文回答也很幫助。非常感謝

**Comment by LEOYoon-Tsaw** [comment]
In `cangjie5.dict.yaml`, there's a section of `encoder/rules`. Those word encoding rules are defined there.
Capital letters locate the character in the word, small letters locate the code to use. While A, B, C count from 1, 2, 3, Z, Y count as -1, -2.

**Reply by erikli** [reply]
thank you!

**Comment by lotem** [comment, marked as answer]
It's using a non-official phrase encoding sheme from 快趣倉頡輸入法.
There were discussions about it on the [倉頡之友 forum](https://www.chinesecj.com/forum/), and many Cangjie enthusiasts prefer this method because it applies similar encoding rules between characters and phrases.

Like treating the internal structure of a Chinese character, a phrase is spit apart into 2 or 3 groups of characters, namely head, (body,) tail.
Attached image shows a few examples.
![20090224_72139de30e9d821218a2TnxQdNbAW4CU.jpeg](https://user-images.githubusercontent.com/151157/173167796-cf4fe5c4-ef6e-482a-b81d-13f17f96e951.jpeg)

Understood the encoding rules, we can implement them in the Rime dictionary as follows.
A for the first character, B for the second, etc. Z for the last character. Lowercase a, b, z refer to the first, second and last Cangjie letter of the character.

```yaml
encoder:
  rules:
    - length_equal: 2
      formula: "AaAzBaBbBz"
    - length_equal: 3
      formula: "AaAzBaBzCz"
    - length_in_range: [4, 10]
      formula: "AaBzCaYzZz"
```

**Reply by erikli** [reply]
This is so helpful, thank you!
